src/indieweb_utils/indieauth/server.py

```diff
@@ -0,0 +1,308 @@
+
+
+def verify_decoded_code(
+        client_id: str,
+        redirect_uri: str,
+        decoded_client_id: str,
+        decoded_redirect_uri: str,
+        decoded_expires: int
+    ) -> bool:
+    """
+        Conducts checks to validate the decoded code in an authorization request.
+
+        :param client_id: The client ID of the authorization request.
+        :rtype: str
+        :param redirect_uri: The redirect URI of the authorization request.
+        :rtype: str
+        :param decoded_client_id: The decoded client ID of the authorization request.
+        :rtype: str
+        :param decoded_redirect_uri: The decoded redirect URI of the authorization request.
+        :rtype: str
+        :param decoded_expires: The decoded expiration time of the authorization request.
+        :rtype: int
+        :returns: True if the decoded code is valid, False otherwise.
+        :rtype: bool
+    """
+
+    if int(time.time()) > decoded_expires:
+        raise AuthorizationCodeExpiredError("The authorization code has expired.")
+
+    if redirect_uri != decoded_redirect_uri:
+        raise TokenValidationError(
+            "The redirect URI provided does not match the redirect URI in the authorization token."
+        )
+
+    if client_id != decoded_client_id:
+        raise TokenValidationError(
+            "The client ID provided does not match the client ID in the authorization token."
+        )
+
+    return True
+
+
+def generate_auth_token(
+        me: str,
+        client_id: str,
+        redirect_uri: str,
+        response_type: str,
+        state: str,
+        code_challenge: str,
+        code_challenge_method: str,
+        final_scope: str,
+        secret_key: str,
+        **kwargs
+    ) -> str:
+    """
+        Generates an IndieAuth authorization token.
+
+        :param me: The URL of the user's profile.
+        :type me: str
+        :param client_id: The client ID of the authorization request.
+        :type client_id: str
+        :param redirect_uri: The redirect URI of the authorization request.
+        :type redirect_uri: str
+        :param response_type: The response type of the authorization request.
+        :type response_type: str
+        :param state: The state of the authorization request.
+        :type state: str
+        :param code_challenge: The code challenge, used for PKCE.
+        :type code_challenge: str
+        :param code_challenge_method: The code challenge method, used for PKCE.
+        :type code_challenge_method: str
+        :param final_scope: The scopes approved by the user.
+        :type final_scope: str
+        :param secret_key: The secret key used to sign the token.
+        :type secret_key: str
+        :param kwargs: Additional parameters to include in the token.
+        :type kwargs: dict
+        :returns: The authorization token.
+        :rtype: str
+    """
+
+    if not client_id or not redirect_uri or not response_type or (not state and state != ""):
```

Comment:
According to the typehints and docstrings we'll always be passing in strings for every parameter, so we could simplify this.

```suggestion
    if all([client_id, redirect_uri,  response_type, state]):
```